### DISCUSSION TITLE
Parallel (thing) changes

### DISCUSSION THREAD

**Discussion by dherkes** [jasoncoon/esp8266-fastled-webserver#218] [Q&A] [answered]
I have compiled and uploaded the newly updated parallel sketch.  I really like the changes made so far, but I have a question.  Where do I go about defining the NUM_LEDS variable?  I have six strings of 50 LEDS.  Thank you for all the hard work and intelligent design

**Comment by jasoncoon** [comment]
I'm still learning the new config layout as well, but I think this is where you'd change `NUM_LEDS` for what was previously the `parallel` branch: https://github.com/jasoncoon/esp8266-fastled-webserver/blob/main/esp8266-fastled-webserver/include/configs/product/esp8266_thing.h#L35

Lower in that same file you'll also find `PARALLEL_OUTPUT_CHANNELS`, `PIXELS_ON_DATA_PIN_X`, and `DATA_PIN_X`.

If you're using Arduino, you'll also need to comment out `#define PRODUCT_DEFAULT`, and uncomment `#define PRODUCT_ESP8266_THING`.

If you haven't already, I highly recommend trying out the new PlatformIO support that @henrygab has worked so hard on adding: https://github.com/jasoncoon/esp8266-fastled-webserver/discussions/207#discussioncomment-1655195

It really is so much easier and better.  I now regret not trying it sooner. 😆

**Reply by dherkes** [reply]
Thanks, Jason.  I worked it out.  I will move to the PlatformIO asap.  I have outgrown the Arduino (but used it for 12 years!)  I can't wait to try my hand at a parallel ESP32 construct.

**Comment by henrygab** [comment, marked as answer]
Your post made me realize I had not added `esp_thing` as a build environemt in PlatformIO.ini.   I'm fixing that now.  :)   The instructions below should work even with current `platformio.ini` in main branch.

Expand either instruction below for details.

<hr/><details><summary>In Arduino</summary><P/>

In the Arduino environment, you simply edit config.h as follows:

https://github.com/jasoncoon/esp8266-fastled-webserver/blob/b1dd8e9938eca5a93f8d11b1a8bf0389d653b3e6/esp8266-fastled-webserver/config.h#L27-L28

You would replace line 28 with the following contents:

```c++
    #define PRODUCT_ESP8266_THING
    #define NUM_PIXELS 72
    #define PIXELS_ON_DATA_PIN_1 5
    #define PIXELS_ON_DATA_PIN_2 7
    #define PIXELS_ON_DATA_PIN_3 11
    #define PIXELS_ON_DATA_PIN_4 13
    #define PIXELS_ON_DATA_PIN_5 17
    #define PIXELS_ON_DATA_PIN_6 19
```

</details><HR/>
<details><summary>In PlatformIO</summary><P/>

PlatformIO allows you to add build environments via the file `platformio_override.ini`.  Just create it in the root of the depot, with the following contents:


```ini

[env:custom_esp_thing_older]
platform = ${common.platform_default}
platform_packages = ${common.platform_packages}
lib_deps = ${esp8266.lib_deps}
board = d1_mini
board_build.ldscript = ${common.ldscript_4m1m}
build_unflags = ${common.build_unflags}
build_flags =
	${common.build_flags_esp8266}
	-D PRODUCT_ESP8266_THING
	-D NUM_PIXELS=72
	-D PIXELS_ON_DATA_PIN_1=5
	-D PIXELS_ON_DATA_PIN_2=7
	-D PIXELS_ON_DATA_PIN_3=11
	-D PIXELS_ON_DATA_PIN_4=13
	-D PIXELS_ON_DATA_PIN_5=17
	-D PIXELS_ON_DATA_PIN_6=19

```

Then, manually refresh the PlatformIO project list:
![image](https://user-images.githubusercontent.com/740168/143660983-8e1cdef4-720f-4b2c-bc15-6647e470903b.png)

That's all.  Your custom build environment will appear, and expanding it will give you the same options to build / upload file system / etc.

</details><HR/>

**Reply by dherkes** [reply]
Thank you so much.

**Comment by henrygab** [comment]
[update] Not critical, but because the default is six strings of 50 pixels, for your particular situation, you don't even have to define `NUM_PIXELS` or any of the `PIXELS_ON_DATA_PIN_x`.
